### PR TITLE
Update-docs-algorithm

### PR DESCRIPTION
**Pull request by tijlleenders**
Updated algorithm to make it more readable. Corrected some minor details.

### CODE REVIEW

**Comment by tijlleenders** [comment]
@thinkrapido FYI

**Comment by kobe-reygel** [comment]
Overall a great improvement of the documentation.

I left some small remarks that were not 100% clear for me

**Comment by tijlleenders** [comment]
> I left some small remarks that were not 100% clear for me. 

@kobe-reygel 
I don't see any remarks / commits.

**Review by kobe-reygel — COMMENTED** [review]
@tijlleenders if I do this, can you see my review comments?

File: documentation/functional/Algorithm-Overview.md [hunk 1]
```diff
@@ -1,51 +1,73 @@
...
-We can also have budgets with a minimum and a maximum amount of time to spend on a goal in a given timeframe
-(e.g. spend between 4 and 8 hours weekly of studying Rust)
-  
-### 0) Add filler Goals if required  
-The Goals are presented in a Directed Acyclic Graph (DAG), like a tree. Only the Goals at the 'bottom of the DAG' will be considered for generating Steps. We call these the 'leaves' of the tree.
+
+## Preparation of Steps and Budgets
+
+### Given: Goals and Budgets
+We start with defined Goals with constraints and modifiers.
```
[the remark this answers is not in the document]
Comment (tijlleenders) [reply documentation/functional/Algorithm-Overview.md:12]: Yeah totally!  

From the top of my head a few constraints/modifiers that impact how Steps and Step-Timelines are created:

For Budgets:
- after/before time (of day)
- on_days
- start
- not_on (not used yet)

For Goals:
- after/before time (of day)
- not_on (not used yet)
- start
- deadline
- constraints from any budget ancestors of the Goal
- repeat interval
- after_goal (can only start after other goal completes)

[the remark this answers is not in the document]
Comment (kobe-reygel) [reply documentation/functional/Algorithm-Overview.md:12]: fantastic, let's also put this into the documentation then.
(I'm willing to do it, but I probably won't have the time today)

File: documentation/functional/Algorithm-Overview.md [hunk 2]
```diff
@@ -1,51 +1,73 @@
...
+e.g. 'run 4 hours every week' generates a Step 'run 4 hours between monday and sunday' for every
 week between the start and end date.
 
-Goals with durations <= 8 hours are kept as a single block. Goals with a longer duration are split into 1-hour sized steps.
+Goals with durations <= 8 hours are kept as a single block. Goals with a longer duration are split into 1-hour sized Steps.  
 
-### 3) Generate slots
-Go over the timeline and - for each step - calculate all the time slots that are available for that step.
+Budgets can also generate Steps - irrespective of their place in the hierarchy. They are only placed after all 'regular' Steps are placed.  
+NB!: Whenever a 'regular' Step belonging to the Budget is placed, any corresponding Budget Steps get removed for the corresponding time periods and durations.
```
[the remark this answers is not in the document]
Comment (tijlleenders) [reply documentation/functional/Algorithm-Overview.md:35]: Let's discuss this in a quick call when you're available. I think we're on the same page - just using different words :)

File: documentation/functional/Algorithm-Overview.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tijlleenders) [reply documentation/functional/Algorithm-Overview.md:12]: @kobe-reygel I think this is done with [commit 1be8b9e](https://github.com/tijlleenders/ZinZen-scheduler/pull/409/commits/1be8b9eee1d2488ff7a78a8665a53fd05e3625a3).  
I found the most appropriate place was not the algorithm documentation - but the core concepts page.

File: documentation/functional/Algorithm-Overview.md [hunk 3]
```diff
@@ -1,51 +1,73 @@
...
-Update flexibility for remaining steps (+ remove slot from those steps).
-Decrease budget if the goal corresponding to the step has one,
+
+### 4) Update Timeline of other Steps where needed  
+Remove the scheduled Slot(s) from the Timeline of any Steps that still had the Slot(s) (or part of it) as 'available'.  
+
+
+### 5) Update Budgets and Budget Steps  
+
+Decrease Budget if the Goal corresponding to the Step has one, and remove any minimum Budget Steps for corresponding period and duration.
```
[the remark this answers is not in the document]
Comment (tijlleenders) [reply documentation/functional/Algorithm-Overview.md:59]: I'll remove this comment and go for optional Budget Steps that are placed last.

**Comment by tijlleenders** [comment]
@kobe-reygel @thinkrapido  
Improved the docs and added an image showing the correct way of scheduling too in 'Algorithm-overview'.  

Can we merge this now and make separate issues for anything else?

File: documentation/functional/Readme.md [hunk 4]
```diff
@@ -12,4 +12,4 @@ describe step-by-step how the scheduling algorithm works.
 The pdf [ZinZen Algorithm Test Case Simple](ZinZen%20Algorithm%20Test%20Case%20Simple.pdf) visually
 represents an algorithm test run scheduling 3 simple goals. The PDF contains 
 a detailed description of each step and is a good place to start understanding
-(a simple case) of the algorithm.
\ No newline at end of file
+(a simple case) of the algorithm. It does contain a small error - can you spot it?
```
Comment (kobe-reygel) [inline documentation/functional/Readme.md:15]: hahha love it

**Review by kobe-reygel — APPROVED** [review]
looks good

**Comment by kobe-reygel** [comment]
I'm not 100% convinced that the fact that we structure goals and budgets in a DAG needs to be part of the funcitonal ubitiquous language: is this really an integral part of the algorithm design, or is this rather an implementation choice?

regardless, looks good I think we can merge it as is for now :D @tijlleenders

**Comment by tijlleenders** [comment]
> I'm not 100% convinced that the fact that we structure goals and budgets in a DAG needs to be part of the funcitonal ubitiquous language: is this really an integral part of the algorithm design, or is this rather an implementation choice?

@kobe-reygel The tree structure is how a front-end user structures 'their world'. A top level budget 'Work' or 'Family time' and subGoals with 'Project A' or 'Customer XYZ' and maybe evel levels below that like 'Shopping' and items per category. It's like a 'memory palace' for goals that also impacts from which budgets or time allocations something comes. In that sense I believe the DAG is part of the user experience / domain.

**Comment by tijlleenders** [comment]
@kobe-reygel 
What we could do is make the relationship explicit and name it - instead of using an implicit structure.  
Good point!

**Comment by kobe-reygel** [comment]
Hm I also see your point. That is good for thought, but in the end that's just a detail, and this is non-blocking :)

**Pull request merged**